### PR TITLE
Tripal 4 ChadoAnalysisDefault field

### PR DESCRIPTION
**Pull request by dsenalik**
# Tripal 4 Core Dev Task

## Description

This is the initial implementation of the ChadoAnalysisDefault field.

There are only two content types with an `analysis_id` column that will be able to use this field:
  * `phylotree`
  * `quantification`

## Testing?
I am testing it on the phylogenetic tree content type, but we don't have a dbxref field yet, so right now you can't manually enter a tree. To be able to test, I have to temporarily do two things to allow me to create a phylotree record:
 1. `ALTER TABLE chado.phylotree ALTER COLUMN dbxref_id DROP NOT NULL;`
 2.  In `tripal_chado/src/Plugin/TripalStorage/ChadoStorage.php` line 1048 I temporarily add a `return;` to disable validation.

Then I can create a tree, and test adding and changing an analysis record.

### CODE REVIEW

**Comment by dsenalik** [comment]
A thing I still need to do is figure out how to log an error if you attempt to use the field on a content type that does not support it, _i.e._ there is no `analysis_id` column in the base table.

That's the same as issue #1440 mentions.

(See below, this is added)

**Comment by dsenalik** [comment]
This is ready for review, I have added base table validation, which might even be redundant because now only valid base tables are presented for selection, i.e. those with a foreign key to analysis.

**Comment by laceysanderson** [comment]
Mw and @dsenalik talked about this field in cofest yesterday.

We agreed that the following very helpful method is generic enough to be added to the Chado Field Type base class:

```php
  /**
   * Return a list of candidate base tables. We only want to
   * present valid tables to the user, which are those with
   * a foreign key to the specified object table.
   * The list of tables is returned in an alphabetized list
   * ready to use in a form select.
   *
   * @param string $object_table
   *   The Chado table being linked to (object).
   */
  protected function getBaseTables($object_table);
```

I will wait to do a code/functional review until the above is completed.

We also talked about whether this field should also handle the linking table variant for analysis linking. We think likely yes but will continue to contemplate. In the meantime this PR still stands as is since ChadoStorage is not ready to support linking tables yet. If we decide to do this in the future, it is easy to add a PR adding to this field to support the other variant.

**Comment by dsenalik** [comment]
I have moved the function to the parent class and tested, it appears to still work correctly 🎯
This is ready for review.

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoAnalysisWidgetDefault.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    $elements = [];
+    $elements['record_id'] = [
+      '#type' => 'value',
+      '#default_value' => $record_id,
+    ];
+    $elements['analysis_id'] = $element + [
+      '#type' => 'select',
+      '#options' => $analyses,
+      '#default_value' => $analysis_id,
+      '#placeholder' => $this->getSetting('placeholder'),
```
Comment (laceysanderson) [inline tripal_chado/src/Plugin/Field/FieldWidget/ChadoAnalysisWidgetDefault.php:66]: Where does the placeholder get set?

**Comment by laceysanderson** [comment]
✅ Code Review looks good except for the one question above.
✅ Added chadostorage automated testing for phylotree + quantification use cases.

Now I just need to do a functional review using the phylotree example :-)

**Comment by dsenalik** [comment]
> Where does the placeholder get set?

Oh my. That is in several other fields also, but I have no clue what it is for. I won't be around much any more this week, but will invesitgate next week.

**Review by laceysanderson — APPROVED** [review]
Added code so this field and the dbxref_id are added automatically.
After this I was able to do a functional test using the phylotree content type and it all worked beautifully!

File: tripal_chado/src/Plugin/Field/FieldWidget/ChadoAnalysisWidgetDefault.php (hunk 1 above)
Comment (laceysanderson) [reply tripal_chado/src/Plugin/Field/FieldWidget/ChadoAnalysisWidgetDefault.php:66]: Resolving this in favour of a new issue to look at this in all fields it occurs in.
https://github.com/tripal/tripal/issues/1600

**Pull request merged**